## Objective

You are a data analyst of a consulting company that provides customer insight regarding multiple ticketing system, such as JIRA and Zoho Desk. Your team gather surveys to customers regarding their ticketing system's performance. Your role in the team is to gather reports regarding customer satisfaction and sentiment analysis into a single dashboard and present your insight.

Analyze the following metrics and other insight you can find in the dataset:

- Survey response rate
- Customer Satisfaction score (CSAT)
- Customer Effort Score (CES)
- Net Promoter Score (NPS)
- Sentiment Analysis



## Data Preparation

In [ ]:
import numpy as np
import pandas as pd
import os

pd.options.display.max_columns = 999
pd.options.display.float_format = "{:.2f}".format

### Access to Drive

Write where you put the data in google drive.

In [ ]:
from google.colab import drive
drive.mount('/content/drive/')



Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


### Read Data

Read the file **assignment_ticket_system_review.csv**

In [ ]:
# Replace 'your_folder/your_file.csv' with the actual path to your file in Google Drive
file_path = '/content/drive/MyDrive/FILE DIBIMBING.ID/Data Analysis/assignment_ticket_system_review (1).csv'

try:
    df_review = pd.read_csv(file_path)
    print("File loaded successfully! Here are the first 5 rows:")
    display(df_review.head())
except FileNotFoundError:
    print(f"Error: The file at '{file_path}' was not found. Please check the path and make sure the file exists in your Google Drive.")
except Exception as e:
    print(f"An error occurred while reading the file: {e}")

File loaded successfully! Here are the first 5 rows:


,id_survey,date_of_survey,ticket_system,overall_rating,customer_service,features,value_for_money,ease_of_use,likelihood_to_recommend,overall_text
0,T_02161,2024-11-20,Zendesk,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,T_00229,2024-10-06,Zendesk,3.00,4.00,3.00,3.00,2.00,6.00,Customer tickets managements
2,T_04527,2024-12-26,Zoho Desk,5.00,5.00,5.00,5.00,5.00,8.00,"After 6 months of using the Zoho desk, we shif..."
3,T_03190,2024-12-08,Zoho Desk,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,T_00644,2024-10-17,Zendesk,5.00,3.00,4.00,5.00,5.00,6.00,Pros:Zendesk has always been one of the go-to ...


The following is the dictionary for the data, survey is only valid if all of the survey questions and text review is not blank (null):

**General Information**
- id_survey: identifier for each survey
- date_of_survey: date of survey taken
- ticket_system: The name of the ticket system being reviewed (e.g. Zoho Desk)

**Survey Questions**
- overall_rating: The overall satisfaction rating given by the reviewer, ranging from 1 to 5
- customer_service: The satisfaction rating for the customer service provided by the ticket system, ranging from 1 to 5.
- features: The satisfaction rating for the features of the ticket system, ranging from 1 to 5
- value_for_money: The satisfaction rating for the value for money provided by the ticket system, ranging from 1 to 5
- ease_of_use: The rating for how easy the ticket system is to use, ranging from 1 to 5
- likelihood_to_recommend: The likelihood that the reviewer would recommend the ticket system to others, ranging from 1 to 10
- overall_text: The full text of the overall review, providing detailed feedback on the ticket system.


In [ ]:
# Check the type of data
df_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1462 entries, 0 to 1461
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id_survey                1462 non-null   object 
 1   date_of_survey           1462 non-null   object 
 2   ticket_system            1462 non-null   object 
 3   overall_rating           787 non-null    float64
 4   customer_service         787 non-null    float64
 5   features                 787 non-null    float64
 6   value_for_money          787 non-null    float64
 7   ease_of_use              787 non-null    float64
 8   likelihood_to_recommend  787 non-null    float64
 9   overall_text             787 non-null    object 
dtypes: float64(6), object(4)
memory usage: 114.3+ KB


### Data Cleansing

Convert the date column into proper date_time format.

In [ ]:
# Convert data type
df_review['date_of_survey'] = pd.to_datetime(df_review['date_of_survey'])

df_review.head()

,id_survey,date_of_survey,ticket_system,overall_rating,customer_service,features,value_for_money,ease_of_use,likelihood_to_recommend,overall_text
0,T_02161,2024-11-20,Zendesk,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,T_00229,2024-10-06,Zendesk,3.00,4.00,3.00,3.00,2.00,6.00,Customer tickets managements
2,T_04527,2024-12-26,Zoho Desk,5.00,5.00,5.00,5.00,5.00,8.00,"After 6 months of using the Zoho desk, we shif..."
3,T_03190,2024-12-08,Zoho Desk,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,T_00644,2024-10-17,Zendesk,5.00,3.00,4.00,5.00,5.00,6.00,Pros:Zendesk has always been one of the go-to ...


In [ ]:
# handling duplicates
len(df_review.drop_duplicates())

1462

In [ ]:
len(df_review)

1462

In [ ]:
# handling missing value
df_review.isnull().sum()

,0
id_survey,0
date_of_survey,0
ticket_system,0
overall_rating,675
customer_service,675
features,675
value_for_money,675
ease_of_use,675
likelihood_to_recommend,675
overall_text,675


In [ ]:
for column in df_review.columns:
    print(f"============= {column} =================")
    display(df_review[column].value_counts())
    print()

============= id_survey =================


,count
id_survey,
T_01981,1
T_02161,1
T_00229,1
T_04527,1
T_03190,1
...,...
T_03971,1
T_00355,1
T_01238,1



============= date_of_survey =================


,count
date_of_survey,
2024-12-10,39
2024-12-12,32
2024-12-19,30
2024-12-29,29
2024-11-18,28
...,...
2024-10-24,8
2024-10-16,7
2024-10-13,7



============= ticket_system =================


,count
ticket_system,
Zoho Desk,438
Zendesk,435
Freshdesk,355
Jira Service Management,144
ServiceNow,54
otrs,36



============= overall_rating =================


,count
overall_rating,
5.00,492
4.00,251
3.00,38
2.00,4
1.00,2



============= customer_service =================


,count
customer_service,
4.00,269
3.00,256
5.00,116
2.00,99
1.00,30
0.00,15
-1.00,2



============= features =================


,count
features,
5.00,419
4.00,293
3.00,61
2.00,12
1.00,2



============= value_for_money =================


,count
value_for_money,
5.00,435
4.00,247
3.00,80
2.00,19
1.00,6



============= ease_of_use =================


,count
ease_of_use,
5.00,466
4.00,246
3.00,60
2.00,12
1.00,3



============= likelihood_to_recommend =================


,count
likelihood_to_recommend,
8.00,224
9.00,174
7.00,155
6.00,82
10.00,77
5.00,34
4.00,18
3.00,10
2.00,5



============= overall_text =================


,count
overall_text,
Pros:Easy to start with a basic free setup and upgrade as your company grows or develops the need for more features.,1
Customer tickets managements,1
"After 6 months of using the Zoho desk, we shifted to different software, but my experience using Zoho was great! Indeed, I would still recommend using Zoho after all.",1
"Pros:Zoho Desk is Ease to use,Customer Service.",1
Pros:I enjoy the UI. It's simple and intuitive.,1
...,...
"Served us incredibly well for years, the macros and auto-responders cut down a lot of work for our CS team. It allowed us to add a cost-effective,personalised voice mail, call recording and a comprehensive help desk to support our customers.",1
"Its best, but need extra improvement from UI Side",1
Pros:We have connected Zendesk with a lot of our other applications.,1


In [ ]:
# Calculate the percentage of missing values
missing_percentage = df_review.isnull().sum() / len(df_review) * 100

# Display the percentage of missing values
print("Percentage of missing values in each column:")
print(missing_percentage)

Percentage of missing values in each column:
id_survey                  0.00
date_of_survey             0.00
ticket_system              0.00
overall_rating            46.17
customer_service          46.17
features                  46.17
value_for_money           46.17
ease_of_use               46.17
likelihood_to_recommend   46.17
overall_text              46.17
dtype: float64


Handling missing value dilakukan di akhir setelah analisis dilakukan agar tidak berpengaruh pada hasil analis seperti pada column overal rating yang NaN

## Survey Analysis

### Response Rate

Start by analyzing how many customers has filled the survey, indicated by whether the overall_rating is not blank.

In [ ]:
# How many customer responded to the survey?
df_review['fill_survey'] = np.where(df_review['overall_rating'].isnull(), 'Not Responded', 'Responded')

df_review.value_counts('fill_survey', normalize = True).reset_index()

,fill_survey,proportion
0,Responded,0.54
1,Not Responded,0.46


Create a new dataframe that consists only of those who have responded the survey to simplify calculating the CSAT, CES, and NPS Score.

In [ ]:
# Responded Customer
responded_customer = df_review[df_review['fill_survey'] == 'Responded'].copy()

### CSAT Score

Measure the customer's overall satisfaction score (CSAT) with the following formula:

$$
CSAT = \frac{\Sigma\ total\ satisfaction\ score}{number\ of\ responded\ customer \times \max\ rating}
$$

The max rating is inserted to convert the CSAT score into percentage.

CSAT score can be classified into categories based on the result. There is no absolute threshold for each categories but the following is the common threshold:

- \>= 90%: Excellent
- 75%-90%: Good
- 60-75%: Fair
- \<60%: Poor

In [ ]:
# CSAT Score
max_rating = 5
# Menghitung nilai maksimum total rating yang diperoleh dari seluruh pelanggan yang merespon
n_data = (responded_customer.shape[0] * max_rating)

# Menjumlahkan semua rating dan kemudian menghitung skor CSAT
csat_score = responded_customer['overall_rating'].sum() / n_data

print(f'Overal CSAT Score: { (csat_score * 100):.1f}%')

Overal CSAT Score: 91.2%


Berdasarkan skor CSAT sebesar 91,2%, tingkat kepuasan pelanggan secara keseluruhan dinilai Sangat Baik. Hal ini menunjukkan bahwa sebagian besar pelanggan sangat puas dengan sistem tiket.

Measure the satisfaction score for the following attributes:

- customer service
- features
- value for money

In [ ]:
# Satisfaction Score for Attributes

customer_service_csat = responded_customer['customer_service'].sum() / n_data
features_csat = responded_customer['features'].sum() / n_data
value_for_money_csat = responded_customer['value_for_money'].sum() / n_data

print(f'Overal CSAT Score: { (csat_score * 100):.1f}%')
print(f'Customer Service CSAT Score: { (customer_service_csat * 100):.1f}%')
print(f'Features CSAT Score: { (features_csat * 100):.1f}%')
print(f'Value for Money CSAT Score: { (value_for_money_csat * 100):.1f}%')

Overal CSAT Score: 91.2%
Customer Service CSAT Score: 67.3%
Features CSAT Score: 88.3%
Value for Money CSAT Score: 87.6%


Customer Service CSAT Score: 67,3% masuk dalam kategori Cukup. Meskipun kepuasan secara keseluruhan sangat baik,Customer Service merupakan area yang dapat ditingkatkan.
Skor CSAT Fitur: 88,3% termasuk dalam kategori Baik, mendekati Sangat Baik. Pelanggan umumnya puas dengan fitur yang disediakan oleh sistem tiket.
Value for Money CSAT Score: 87,6% termasuk dalam kategori Baik. Pelanggan menganggap sistem tiket menawarkan nilai yang baik untuk uang mereka.

### CES Score

Measure CES with the following formula


$$
CES = \frac{\Sigma\ total\ effort\ score}{number\ of\ responded\ customer \times \max\ rating}
$$

In [ ]:
# CES Score
max_rating = 5

ces_ease_of_use = responded_customer['ease_of_use'].sum() / n_data

print(f'Ease of Use CES Score: { (ces_ease_of_use * 100):.1f}%')

Ease of Use CES Score: 89.5%


 Ease of Use CES Score: 89.5% menunjukkan bahwa pelanggan umumnya menemukan sistem tiket mudah digunakan.

### NPS Score

To calculate the NPS score, first we must convert the **would_you_recommend** column into proper NPS Category based on the rating value:

* Promoter: Rating 9-10
* Passive: Rating 7-8
* Detractor: Rating < 7

In [ ]:
# Category NPS
nps_value = ['promoter', 'passive', 'detractor']
nps_condition = [responded_customer['likelihood_to_recommend'].isin([9, 10]),
                 responded_customer['likelihood_to_recommend'].isin([7, 8]),
                 responded_customer['likelihood_to_recommend'] < 7]

responded_customer['nps_category'] = np.select(nps_condition, nps_value, default='Unknown')

responded_customer.value_counts('nps_category', normalize = True)

,proportion
nps_category,
passive,0.48
promoter,0.32
detractor,0.20



Calculate the NPS Score with the following formula

$$
NPS = \frac{Promoter - Detractor}{Total\ Survey\ Responded}
$$

In [ ]:
# NPS Score
nps_agg_1 = responded_customer.value_counts('nps_category').reset_index()

nps_agg_promoter = nps_agg_1[nps_agg_1['nps_category'] == 'promoter'] ['count'].item()
nps_agg_detractor = nps_agg_1[nps_agg_1['nps_category'] == 'detractor'] ['count'].item()
nps_score = (nps_agg_promoter - nps_agg_detractor) / nps_agg_1['count'].sum()

print(f'NPS Score: { (nps_score * 100):.1f}')

NPS Score: 11.9


Berdasarkan skor NPS yang dihitung sebesar 11,9, skor Net Promoter Score (NPS) termasuk dalam kategori Rata-rata dan Detractor 20% Pelanggan ini tidak puas dan kemungkinan besar akan menyebarkan hal negatif.

NPS score can be ranging from -100 (when all customers are detractor) to 100 (when all customers are promoter).

NPS Score can be classified into categories based on the following threshold:

- \>= 70: Excellent
- 50-69: Very Good
- 30-49: Good
- 0-29: Average
- \< 0: Poor

## Sentiment Analysis

Create a new dataframe with no blank overall_text.

In [ ]:
# Create new dataframe
df_review_clean = df_review.dropna(axis =0, subset = 'overall_text')[['id_survey', 'overall_text']].copy()

df_review_clean.head()

,id_survey,overall_text
1,T_00229,Customer tickets managements
2,T_04527,"After 6 months of using the Zoho desk, we shif..."
4,T_00644,Pros:Zendesk has always been one of the go-to ...
6,T_04682,It has been very useful so far to integrate mu...
8,T_01238,Pros:It's easy to use and very intuitive.We ha...


### Text Cleansing

In order to get more accurate sentiment, several text cleansing need to be done. However, in most of recent sentiment analysis models and algorithm, the only text cleansing needed are as follows:

* Clean double whitespace
* Clean URL/website
* Clean username (mostly in social media or digital text)

In [ ]:
import re

def cleansing_text(x):
  # clean double whitespace
  out_text = ' '.join(x.split())

  # clean url
  out_text = re.sub(r"http\S+|www\S+|https\S+", 'http', out_text)

  # clean username
  out_text = re.sub(r"@\S+", '@user', out_text)

  return(out_text)

cleansing_text(" Doesn't  dissapoint. The car       was great. It was the best car rental experiences I've had! Salute to @jone who recommend https:/rental.com")

"Doesn't dissapoint. The car was great. It was the best car rental experiences I've had! Salute to @user who recommend http"

In [ ]:
# apply cleansing to review
df_review_clean['clean_text'] = df_review_clean['overall_text'].apply(cleansing_text)

df_review_clean.head()

,id_survey,overall_text,clean_text
1,T_00229,Customer tickets managements,Customer tickets managements
2,T_04527,"After 6 months of using the Zoho desk, we shif...","After 6 months of using the Zoho desk, we shif..."
4,T_00644,Pros:Zendesk has always been one of the go-to ...,Pros:Zendesk has always been one of the go-to ...
6,T_04682,It has been very useful so far to integrate mu...,It has been very useful so far to integrate mu...
8,T_01238,Pros:It's easy to use and very intuitive.We ha...,Pros:It's easy to use and very intuitive.We ha...


### Sentiment Analysis

Create a sentiment categories using algorithm of your own choice.

In [ ]:
import nltk
from nltk.sentiment.vader import SentimentIntensityAnalyzer

# Download VADER
try:
    analyzer = SentimentIntensityAnalyzer()
except LookupError:
    nltk.download('vader_lexicon')
    analyzer = SentimentIntensityAnalyzer()


# Create sentiment category
def categorize_sentiment(text):
    score = analyzer.polarity_scores(text)['compound']
    if score >= 0.05:
        return 'Positive'
    elif score <= -0.05:
        return 'Negative'
    else:
        return 'Neutral'

# menerapkan fungsi kategorisasi pada teks yang telah dibersihkan.
df_review_clean['sentiment'] = df_review_clean['clean_text'].apply(categorize_sentiment)

df_review_clean.head()

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...


,id_survey,overall_text,clean_text,sentiment
1,T_00229,Customer tickets managements,Customer tickets managements,Neutral
2,T_04527,"After 6 months of using the Zoho desk, we shif...","After 6 months of using the Zoho desk, we shif...",Positive
4,T_00644,Pros:Zendesk has always been one of the go-to ...,Pros:Zendesk has always been one of the go-to ...,Positive
6,T_04682,It has been very useful so far to integrate mu...,It has been very useful so far to integrate mu...,Positive
8,T_01238,Pros:It's easy to use and very intuitive.We ha...,Pros:It's easy to use and very intuitive.We ha...,Positive


Check the number of data by sentiment.

In [ ]:
# Number of Sentiment
df_review_clean.value_counts('sentiment')

,count
sentiment,
Positive,673
Neutral,86
Negative,28


## Finalize Data for Reporting

Save the review data with NPS category and sentiment information to new csv for the dashboard.

In [ ]:
# Save Data
df_review_final = df_review.merge(df_review_clean[['id_survey', 'sentiment']],
                                  on = 'id_survey', how = 'left'
                                  )

df_review_final.head()

,id_survey,date_of_survey,ticket_system,overall_rating,customer_service,features,value_for_money,ease_of_use,likelihood_to_recommend,overall_text,fill_survey,sentiment
0,T_02161,2024-11-20,Zendesk,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Not Responded,NaN
1,T_00229,2024-10-06,Zendesk,3.00,4.00,3.00,3.00,2.00,6.00,Customer tickets managements,Responded,Neutral
2,T_04527,2024-12-26,Zoho Desk,5.00,5.00,5.00,5.00,5.00,8.00,"After 6 months of using the Zoho desk, we shif...",Responded,Positive
3,T_03190,2024-12-08,Zoho Desk,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Not Responded,NaN
4,T_00644,2024-10-17,Zendesk,5.00,3.00,4.00,5.00,5.00,6.00,Pros:Zendesk has always been one of the go-to ...,Responded,Positive


In [ ]:
# Drop baris yang mengandung nilai NaN pada kolom survei utama dan sentimen.
survey_and_sentiment_cols = ['overall_rating', 'customer_service', 'features', 'value_for_money', 'ease_of_use', 'likelihood_to_recommend', 'overall_text', 'sentiment']
df_review_final_cleaned = df_review_final.dropna(subset=survey_and_sentiment_cols).copy()

# Display  jumlah baris setelah menghapus nilai yang hilang.
print(f"Number of rows after dropping missing values in df_review_final: {len(df_review_final_cleaned)}")

# Display  jumlah nilai yang hilang di setiap kolom setelah menghapus.
print("\nMissing values after dropping rows in df_review_final:")
print(df_review_final_cleaned.isnull().sum())

Number of rows after dropping missing values in df_review_final: 787

Missing values after dropping rows in df_review_final:
id_survey                  0
date_of_survey             0
ticket_system              0
overall_rating             0
customer_service           0
features                   0
value_for_money            0
ease_of_use                0
likelihood_to_recommend    0
overall_text               0
fill_survey                0
sentiment                  0
dtype: int64


In [ ]:
df_review_final_cleaned.head()

,id_survey,date_of_survey,ticket_system,overall_rating,customer_service,features,value_for_money,ease_of_use,likelihood_to_recommend,overall_text,fill_survey,sentiment
1,T_00229,2024-10-06,Zendesk,3.00,4.00,3.00,3.00,2.00,6.00,Customer tickets managements,Responded,Neutral
2,T_04527,2024-12-26,Zoho Desk,5.00,5.00,5.00,5.00,5.00,8.00,"After 6 months of using the Zoho desk, we shif...",Responded,Positive
4,T_00644,2024-10-17,Zendesk,5.00,3.00,4.00,5.00,5.00,6.00,Pros:Zendesk has always been one of the go-to ...,Responded,Positive
6,T_04682,2024-12-28,Zoho Desk,5.00,4.00,5.00,5.00,5.00,8.00,It has been very useful so far to integrate mu...,Responded,Positive
8,T_01238,2024-11-02,Freshdesk,4.00,4.00,4.00,5.00,4.00,8.00,Pros:It's easy to use and very intuitive.We ha...,Responded,Positive


In [ ]:
df_review['overall_rating'].mean()

np.float64(4.559085133418043)

In [ ]:
# Specify the output file path
output_file_path = '/content/assignment_ticket_system_review_cleaned.csv'

# Save the DataFrame to a CSV file
df_review_final_cleaned.to_csv(output_file_path, index=False)